exporter/jsontypeexporter/exporter.go

```diff
@@ -0,0 +1,333 @@
+	maskArrayFloat   uint16 = 1 << 6
+	maskArrayInt     uint16 = 1 << 7
+	maskArrayString  uint16 = 1 << 8
+	maskArrayJSON    uint16 = 1 << 9
+)
+
+// api.parameters.list.search -> maps are flattened
+//
+// api.routes:kubernetes.container_name -> : is used as nestedness indicator in Arrays
+//
+// analyzePValue walks OTel pcommon.Value without converting to Go maps/slices, minimizing allocations.
+func (e *jsonTypeExporter) analyzePValue(ctx context.Context, prefix string, inArray bool, val pcommon.Value, typeSet *TypeSet) error {
+	switch val.Type() {
+	case pcommon.ValueTypeMap:
+		m := val.Map()
+		m.Range(func(key string, value pcommon.Value) bool {
+			select {
+			case <-ctx.Done():
+				return false
+			default:
+			}
+
+			contains := e.keyCache.Contains(key)
+			// if high cardinality, skip the key
+			if !contains && keycheck.IsCardinal(key) {
+				return true
+			}
+			if !contains {
+				e.keyCache.Add(key, struct{}{}) // add key to cache to avoid checking it again for next log records
+			}
+
+			path := prefix + "." + key
+			if prefix == "" {
+				path = key
+			} else if inArray {
+				path = prefix + key
+			}
+			if err := e.analyzePValue(ctx, path, false, value, typeSet); err != nil {
+				return false
+			}
+
+			return true
+		})
+		return nil
+	case pcommon.ValueTypeSlice:
+		s := val.Slice()
+		var prev uint16
+		mixed := false
+		for i := 0; i < s.Len(); i++ {
+			el := s.At(i)
+			var cur uint16
+			switch el.Type() {
+			case pcommon.ValueTypeMap:
+				// analyze first object deeply for path discovery
+				if err := e.analyzePValue(ctx, prefix+":", true, el, typeSet); err != nil {
+					return err
+				}
+				cur = maskArrayJSON
+			case pcommon.ValueTypeSlice:
+				return fmt.Errorf("arrays inside arrays are not supported! found at path: %s", prefix)
+			case pcommon.ValueTypeStr:
+				cur = maskArrayString
+			case pcommon.ValueTypeBool:
+				cur = maskArrayBool
+			case pcommon.ValueTypeDouble:
+				cur = maskArrayFloat
+			case pcommon.ValueTypeInt:
+				cur = maskArrayInt
+			default:
+				return fmt.Errorf("unknown element type in array at path: %s", prefix)
+			}
+			if i > 0 && cur != prev {
+				mixed = true
+				break
+			}
+			prev = cur
+		}
+		if mixed {
+			typeSet.Insert(prefix, maskArrayDynamic)
+		} else if prev != 0 {
+			typeSet.Insert(prefix, prev)
+		}
+		return nil
+	case pcommon.ValueTypeStr, pcommon.ValueTypeBytes:
+		typeSet.Insert(prefix, maskString)
+		return nil
+	case pcommon.ValueTypeBool:
+		typeSet.Insert(prefix, maskBool)
+		return nil
+	case pcommon.ValueTypeDouble:
+		typeSet.Insert(prefix, maskFloat)
+		return nil
+	case pcommon.ValueTypeInt:
+		typeSet.Insert(prefix, maskInt)
+		return nil
+	default:
+		return fmt.Errorf("unknown type at path: %s", prefix)
+	}
+}
+
+// persistTypes writes the collected types to the ClickHouse database
+func (e *jsonTypeExporter) persistTypes(ctx context.Context, typeSet *TypeSet, flush bool) error {
+	if !flush && typeSet.counter.Load() < 10_000 {
```

Comment:
this means if I send a few log lines and then stop sending it won't be written ?

shouldn't it be a independent go routine ?

Reply:
removed this